src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
+                    : $"Lease failed to be acquired after election. Winner instance is {winnerInstanceId.Value}.";
+                _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                    _options.Value.Priority, reason));
+                return null;
+            }
 
-                if (updateResult.Result)
+            if (lease == null)
+            {
+                var createTime = ServerDateTime.UtcNow;
+                var interval = _slottedInterval.Calculate(createTime, _options.Value.LeaseInterval);
+                lease = new CosmosDbLease
                 {
-                    lease = updateResult.Lease;
-
-                    // backoff to allow other worker lease instances to update the lease
-                    var delay = _allocationDelay.Calculate(_options.Value.Priority, _options.Value.LeaseInterval);
-
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation delayed [{delay}]. Allocated Lease Id: [{lease.Id}]"));
-
-                    await Task.Delay(delay).ConfigureAwait(false);
-
-                    // activate
-                    var now = ServerDateTime.UtcNow;
-                    lease.Interval = _slottedInterval.Calculate(ServerDateTime.UtcNow, _options.Value.LeaseInterval);
-                    lease.LeasedUntil = now.Add(lease.Interval.Value);
-
-                    updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+                    InstanceId = instanceId,
+                    Interval = interval,
+                    LeasedUntil = createTime.Add(interval),
+                    Priority = _options.Value.Priority,
+                    LeaseType = _options.Value.WorkerType
+                };
+                var createResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryCreateLeaseAsync).ConfigureAwait(false);
+                if (createResult.Result || createResult.Lease==null)
+                {
+                    return createResult.Lease;
+                }
 
-                    if (updateResult.Result)
-                    {
-                        _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                            _options.Value.Priority,
-                            $"Lease acquired. {GenerateLeaseInfo(updateResult.Lease)}"));
+                lease = createResult.Lease;
+            }
 
-                        return updateResult.Lease;
-                    }
 
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation failed lease already acquired. Allocated Lease Id: [{lease.Id}]"));
-                }
+            lease.Priority = _options.Value.Priority;
```

Comment:
Why do we need the code below?

Reply:
The election was won and lease should have been created/updated above.

Reply:
this was added only to handle the exceptional case when despite of the delay not all wp see the same winner. 
Say for example there are 2 wp, wp1 and wp2
1) wp1 query lease requests and elect itself as a winner after delay
2) wp2 comes later add its requests and when queries lease requests he finds he's the winner, but a lease was already created by wp1 having lower priority

This is meant to be considere an edge case that should not happen normally

Reply:
Agreed:
- lease is either updated or created, depending on document id existence
- we use either a single `persisteLease` method or we keep `create` and `update` separate
- if `persistLease` is used, it should:
-- create or update the existing record, depending if id is provided
-- if opeartion succeeds, return persisted lease (with id)
-- if that operation fails, raise extra logging as this is unexpected right now. In future, with shared instanceId, it has to read existing record (based on workerType) and compare with the lease that was to be saved: instanceId, workerType, priority, leasedUntil
-- if the same, `persisteLease` was successful and we return the read lease
-- if different, persist fails and additional logging should be made as this is rather unexpected behaviour

Reply:
- if two methods are used, then the above logic will reside in allocateLease method.

Reply:
above logic happens in the TryPersistLease method in the leaseallocator:
if lease has an Id a create will be performed, an update otherwise. If it fails for any reason a new read will happen to retrieve the current lease, that will be used for telemetry purposes only. In the future also to determine if the lease had same instanceId